test/integration/controllermanager/credentialsbinding/referencecleaner/referencecleaner_test.go

```diff
@@ -0,0 +1,327 @@
+				Type: providerType,
+			},
+			CredentialsRef: corev1.ObjectReference{
+				APIVersion: securityv1alpha1.SchemeGroupVersion.String(),
+				Kind:       "WorkloadIdentity",
+				Name:       workloadIdentity.Name,
+				Namespace:  workloadIdentity.Namespace,
+			},
+		}
+	})
+
+	Context("Credentials of type Secret", func() {
+		JustBeforeEach(func() {
+			// Create the credentialsbinding first and wait until the manager's cache has observed it.
+			// Otherwise, the controller might clean up the Secret/WorkloadIdentities too early because it thinks all referencing CredentialsBindings
+			// are gone. Similar to https://github.com/gardener/gardener/issues/6486
+			By("Create CredentialsBinding")
+			Expect(testClient.Create(ctx, credentialsBinding1)).To(Succeed())
+			log.Info("Created CredentialsBinding for test", "credentialsBinding", client.ObjectKeyFromObject(credentialsBinding1))
+
+			By("Wait until manager has observed the CredentialsBinding")
+			Eventually(func() error {
+				return mgrClient.Get(ctx, client.ObjectKeyFromObject(credentialsBinding1), &securityv1alpha1.CredentialsBinding{})
+			}).Should(Succeed())
+
+			By("Create Secret")
+			Expect(testClient.Create(ctx, secret)).To(Succeed())
+			log.Info("Created Secret for test", "secret", client.ObjectKeyFromObject(secret))
+
+			DeferCleanup(func() {
+				By("Delete CredentialsBinding")
+				Expect(testClient.Delete(ctx, credentialsBinding1)).To(Or(Succeed(), BeNotFoundError()))
+				Eventually(func() error {
+					return testClient.Get(ctx, client.ObjectKeyFromObject(credentialsBinding1), credentialsBinding1)
+				}).Should(BeNotFoundError())
+
+				By("Delete Secret")
+				Expect(testClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)).To(Or(Succeed(), BeNotFoundError()))
+				Expect(controllerutils.RemoveFinalizers(ctx, testClient, secret, "gardener.cloud/gardener")).To(Or(Succeed(), BeNotFoundError()))
+				Expect(testClient.Delete(ctx, secret)).To(Or(Succeed(), BeNotFoundError()))
+				Eventually(func() error {
+					return testClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)
+				}).Should(BeNotFoundError())
+			})
+		})
+
+		Context("no CredentialsBinding referencing the Secret", func() {
+			It("should release the Secret", func() {
+				By("Ensure that Secret is not being deleted")
+				Consistently(func(g Gomega) {
+					g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)).To(Succeed())
+				}).Should(Succeed())
+
+				By("Delete the CredentialsBinding")
+				Expect(testClient.Delete(ctx, credentialsBinding1)).To(Succeed())
+
+				By("Ensure finalizer and labels are removed from Secret")
+				Eventually(func(g Gomega) {
+					g.Expect(testClient.Get(ctx, client.ObjectKeyFromObject(secret), secret)).To(Succeed())
+					g.Expect(secret.Finalizers).To(BeEmpty())
+					g.Expect(secret.Labels).ToNot(And(
+						HaveKey("provider.shoot.gardener.cloud/"+providerType),
+						HaveKey("reference.gardener.cloud/credentialsbinding"),
+					))
```

Comment:
It should test both labels are removed, right?
```suggestion
					g.Expect(secret.Labels).To(And(
						Not(HaveKey("provider.shoot.gardener.cloud/"+providerType)),
						Not(HaveKey("reference.gardener.cloud/credentialsbinding")),
					))
```